### ISSUE TITLE
<random>: uniform_real_distribution(a, b) produces outputs outside [a, b)

### ISSUE THREAD

**Issue by denniskb** [microsoft/STL#1074]
**Describe the bug**
> Produces random floating-point values i, uniformly distributed on the interval [a, b)

*[cppreference.com](https://en.cppreference.com/w/cpp/numeric/random/uniform_real_distribution)*

The below code sample demonstrates how a default-constructed `uniform_real_distribution<float>` (a=0, b=1) generates a `1.0f` if provided a random number with all bits set:

**Command-line test case**
```
#include <cassert>
#include <cstdint>
#include <limits>
#include <random>

struct max_rng
{
	using result_type = std::uint32_t;
	result_type min() const { return 0; }
	result_type max() const { return std::numeric_limits<result_type>::max(); }
	result_type operator()() { return max(); }
};

int main()
{
	max_rng rng;
	std::uniform_real_distribution<float> iid;

	assert(iid(rng) < 1.0f);

	return 0;
}
```

**Expected behavior**
Regardless of the input received from the RNG, `uniform_real_distribution` should never produce `b` as the interval is supposed to be right-*exclusive*. GCC passes this test.

**STL version**
Microsoft Visual Studio Community 2019
Version 16.6.4

**Additional context**
The error seems to originate from [`generate_canonical`](https://github.com/microsoft/STL/blob/c10ae01b4d9508eed9d5f059a120ee7223b6ac12/stl/inc/random#L261), which, when passed in the above code, simplifies down to:

```
return (float) UINT_MAX / (UINT_MAX + 1.0f);
```
which, mathematically speaking, *should* be less than 1, but in practice gets rounded to `1.0f` due to limited floating point precision. `generate_canonical` is flawed in other regards: using a normalizing/scaling division leads to non-uniformly distributed values. Both of these issues could be avoided by generating random floats the "canonical" way, which is to set the bits of the mantissa directly (and simply throw away left-over entropy) along the lines of:
```
// @rnd: uniformly distributed 32bit unsigned int.
return (rnd >> 8) / 164233216.0f
```
which not only produces values strictly from [0, 1) for any value of `rnd`, but also makes sure that all values are equally far spaced apart leading to a perfect uniform distribution.

Also tracked by DevCom-110322 and Microsoft-internal VSO-253526 / [AB#253526](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/253526) .

**Comment by fsb4000** [comment]
@BillyONeal commented:

> As an FYI, this is a bug in the standard; there's currently a paper pending to fix it: [WG21-P0952](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0952r0.html)

@statementreply commented:
>his paper proposes a new specification of generate_canonical, but it doesn't appear to me that this paper would address the issue of uniform_real_distribution specification.

**Additional context**
 [VSO-253526](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/253526) | [DevCom-110322](https://developercommunity.visualstudio.com/content/problem/110322/meow.html)

**Comment by denniskb** [comment]
@fsb4000 Thanks for letting me know. So, currently, by definition we are forced to break the standard because it is inconsistent. We can either

- Follow the dictated implementation, breaking the math. constraints, or
- Conform to the math. constraints but deviate from the dictated implementation.

Would it make sense to opt for the 2nd option? Personally, I'd prefer a math. sound implementation over anything else, seeing as we aren't gonna be standard-conform either way.

**Comment by fsb4000** [comment]
@denniskb Yes, I think the second option is good. But I'm just a random person on the Internet, let's wait for an opinion of the  Microsoft employees :)

**Comment by frederick-vs-ja** [comment]
This is related, although not mandatorily, to #4169 (WG21-P0952R2 and LWG-2524).

**Comment by faithandbrave** [comment]
It seems fixed `std::generate_canonical` range. However, `std::uniform_real_distribution` outputs outside range values as yet (v19.40).

**Comment by CaseyCarter** [comment]
> It seems fixed `std::generate_canonical` range. However, `std::uniform_real_distribution` outputs outside range values as yet (v19.40).

`uniform_real_distribution` uses `generate_canonical` under the hood. `generate_canonical` no longer producing `1.0` consequently makes `uniform_real_distribution` not produce its upper bound. The fix [shipped in VS 2022 17.12 preview 1](https://github.com/microsoft/STL/wiki/Changelog#vs-2022-1712-preview-1), so it's not in toolset version 19.40.

**Comment by BenFrantzDale** [comment]
For those landing here wondering the state of affairs: As of now, clang and gcc trunk still fail but MSVC passes: https://godbolt.org/z/3YK6r5a19
